File: common/src/main/java/org/tdf/common/store/PrefixStore.java
package org.tdf.common.store;

import com.github.salpadding.rlpstream.Rlp;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.tdf.common.serialize.Codec;
import org.tdf.common.util.ByteUtil;
import org.tdf.common.util.HexBytes;
import org.tdf.common.util.RLPUtil;

import java.util.*;


@RequiredArgsConstructor
public class PrefixStore<K, V> implements IterableStore<K, V> {
    private final Store<HexBytes, HexBytes> contractStorage;
    private final HexBytes prefix;
    private final Codec<K> kCodec;
    private final Codec<V> vCodec;

    @Override
    public V get(@NonNull K k) {
        byte[] encoded = kCodec.getEncoder().apply(k);
        HexBytes withPrefix = verifyAndPrefix(encoded);
        HexBytes v = contractStorage.get(
            withPrefix
        );
        return (v == null || v.size() == 0) ? null : vCodec.getDecoder().apply(v.getBytes());
    }

    private HexBytes verifyAndPrefix(byte[] key) {
        if (key.length == 0)
            throw new RuntimeException("invalid key, length = 0");
        return HexBytes.fromBytes(ByteUtil.merge(prefix.getBytes(), key));
    }

    @Override
    public void set(@NonNull K k, @NonNull V v) {
        byte[] encoded = kCodec.getEncoder().apply(k);
        HexBytes withPrefix = verifyAndPrefix(encoded);
        addKey(encoded);
        contractStorage.set(withPrefix, HexBytes.fromBytes(vCodec.getEncoder().apply(v)));
    }

    private TreeSet<HexBytes> keySet() {
        HexBytes keys = contractStorage.get(prefix);
        TreeSet<HexBytes> keySet = keys == null || keys.size() == 0 ?
            new TreeSet<>() :
            new TreeSet<>(
                Arrays.asList(RLPUtil.decode(keys, HexBytes[].class))
            );
        return keySet;
    }

    private void addKey(byte[] key) {
        TreeSet<HexBytes> keySet = keySet();
        keySet.add(HexBytes.fromBytes(key));
        contractStorage.set(prefix, HexBytes.fromBytes(Rlp.encode(keySet.toArray())));
    }

    private void removeKey(byte[] key) {
        TreeSet<HexBytes> keySet = keySet();
        keySet.remove(HexBytes.fromBytes(key));
        contractStorage.set(prefix, HexBytes.fromBytes(Rlp.encode(keySet.toArray())));
    }

    @Override
    public void remove(@NonNull K k) {
        byte[] encoded = kCodec.getEncoder().apply(k);
        HexBytes withPrefix = verifyAndPrefix(encoded);
        removeKey(encoded);
        contractStorage.remove(withPrefix);
    }

    @Override
    public void flush() {
        contractStorage.flush();
    }

    @Override
    public Iterator<Map.Entry<K, V>> iterator() {
        Map<K, V> map = new HashMap<>();
        for (HexBytes key : keySet()) {
            map.put(
                kCodec.getDecoder().apply(key.getBytes()),
                vCodec.getDecoder().apply(
                    contractStorage.get(
                        verifyAndPrefix(key.getBytes())
                    ).getBytes()
                )
            );
        }
        return map.entrySet().iterator();
    }
}

File: common/src/main/java/org/tdf/common/store/PrefixStore.kt
package org.tdf.common.store

import com.github.salpadding.rlpstream.Rlp
import org.tdf.common.serialize.Codec
import org.tdf.common.util.HexBytes
import org.tdf.common.util.decode
import org.tdf.common.util.hex
import java.util.*

class PrefixStore<K, V>(
    private val contractStorage: Store<HexBytes, HexBytes>,
    private val prefix: HexBytes,
    private val kCodec: Codec<K>,
    private val vCodec: Codec<V>
) : IterableStore<K, V> {
    override fun get(k: K): V? {
        val encoded = kCodec.encoder.apply(k)
        val withPrefix = verifyAndPrefix(encoded)
        val v = contractStorage[withPrefix]
        return if (v == null || v.isEmpty()) null else vCodec.decoder.apply(v.bytes)
    }

    private fun verifyAndPrefix(key: ByteArray): HexBytes {
        if (key.isEmpty()) throw RuntimeException("invalid key, length = 0")
        return (prefix.bytes + key).hex()
    }

    override fun set(k: K, v: V) {
        val encoded = kCodec.encoder.apply(k)
        val withPrefix = verifyAndPrefix(encoded)
        addKey(encoded)
        contractStorage[withPrefix] = vCodec.encoder.apply(v).hex()
    }

    private fun keySet(): TreeSet<HexBytes> {
        val keys = contractStorage[prefix]
        return if (keys == null || keys.size == 0) TreeSet() else TreeSet(
            keys.decode(
                Array<HexBytes>::class.java
            ).toList()
        )
    }

    private fun addKey(key: ByteArray) {
        val keySet = keySet()
        keySet.add(key.hex())
        contractStorage[prefix] = Rlp.encode(keySet.toTypedArray()).hex()
    }

    private fun removeKey(key: ByteArray) {
        val keySet = keySet()
        keySet.remove(key.hex())
        contractStorage[prefix] = Rlp.encode(keySet.toTypedArray()).hex()
    }

    override fun remove(k: K) {
        val encoded = kCodec.encoder.apply(k)
        val withPrefix = verifyAndPrefix(encoded)
        removeKey(encoded)
        contractStorage.remove(withPrefix)
    }

    override fun flush() {
        contractStorage.flush()
    }

    override fun iterator(): Iterator<Map.Entry<K, V>> {
        val map: MutableMap<K, V> = mutableMapOf()
        for (key in keySet()) {
            map[kCodec.decoder.apply(key.bytes)] = vCodec.decoder.apply(
                contractStorage[verifyAndPrefix(key.bytes)]!!.bytes
            )
        }
        return map.entries.iterator()
    }
}